tools/tuner.py

```diff
@@ -0,0 +1,603 @@
+        else:
+            return list(range(len(self.flatten_candidates[candidate_idx + 1])))
+
+
+class Tuner(ABC):
+    """
+    Class for creating different configs and choose the config with best perf
+    """
+    DEFAULT_BATCH_SIZE = 50
+    DEFAULT_EARLY_STOP = -1
+    DEFAULT_TIMEOUT = -1
+    DEFAULT_MAX_ITERS = sys.maxsize
+
+    def __init__(
+        self,
+        batch_executor,
+        tunning_space: TuningSpace,
+        batch_size=DEFAULT_BATCH_SIZE,
+        early_stop=DEFAULT_EARLY_STOP,
+        checkpoint="",
+    ):
+        self.batch_executor = batch_executor
+        self.batch_size = batch_size
+        self.early_stop = early_stop
+        self.best_cost = sys.float_info.max
+        self.best = []
+        self.iter = 0
+        self.last_update_iter = 0
+        self.skipped_num = 0
+        self.tunning_space = tunning_space
+        self.checkpoint = checkpoint
+        if self.checkpoint:
+            os.makedirs(os.path.dirname(self.checkpoint), exist_ok=True)
+        assert len(tunning_space.graph_config), "There are no tunable ops"
+
+    def tuner_update(self, config_indices_batch: List[List[int]], costs: List[float]):
+        """
+        Update after each batch of configs was executed
+        """
+        if min(costs) < self.best_cost:
+            self.best_cost = min(costs)
+            self.best = config_indices_batch[costs.index(min(costs))]
+        if self.checkpoint:
+            self.save_status()
+
+    @abstractmethod
+    def get_next_config_indices_batch(self) -> List[List[int]]:
+        """
+        Get the next batch of config indices
+        """
+        pass
+
+    @abstractmethod
+    def load_status(self):
+        """
+        Load the Tuner status from the checkpoint
+        """
+        pass
+
+    @abstractmethod
+    def save_status(self):
+        """
+        Save the Tuner status to the checkpoint
+        """
+        pass
+
+    def tuner_finish(self, tuning_time):
+        """
+        Execute when tuning is finished
+        """
+        print("Tuning ends in", tuning_time, "s")
+        best_config = self.tunning_space.make_config_from_indexes(self.best)
+        print("Best cost:", self.best_cost, "ms")
+        print("Best config:", best_config)
+        utils.attach_configs_to_ir(self.tunning_space.initial_ir, best_config)
+        print(
+            "mlir:\n",
+            self.tunning_space.initial_ir,
+        )
+
+    def run(self, max_iter: int = DEFAULT_MAX_ITERS, timeout: int = DEFAULT_TIMEOUT):
+        """
+        Start of tuning process
+        """
+        if self.early_stop > 0 and self.iter - self.last_update_iter > self.early_stop:
+            # in case of resuming from a saved state and it has already
+            # early-stopped
+            print("Early stop now")
+            return
+        start_time = time.time()
+        spaces_size = self.tunning_space.space_size
+        while self.iter < max_iter and self.iter < spaces_size:
```

Comment:
Do we have plan to support parallelized tuning?

Reply:
yes, It's on the to-do list, will add that after integrated into benchgc